include/dali/core/per_stream_pool.h

```diff
@@ -0,0 +1,207 @@
+ * RAII wrapper. When the wrapper goes out of scope, the object is returned to the pool
+ * and will be available for lease again when the stream finishes the work that's been scheduled
+ * until the lease ended.
+ * Additionally, if `reuse_pending_on_same_stream` flag is set, subsequent requests for an object
+ * specifying the same stream will return that object even if the associated work has not yet
+ * completed - this is useful for objects which are safe for immediate reuse on the same stream,
+ * e.g. device buffers.
+ * If there are no available objects for current device, a new object is created and leased
+ * immediately - it will be returned to the pool when the lease is over.
+ *
+ * Example:
+ * ```
+ * class MyClass {
+ *  public:
+ *   PerStreamPool<GPUWorker> workers;
+ *
+ *   void DoSomeJob(cudaStream_t stream) {
+ *     auto worker = workers.Get(stream);
+ *     worker->DoYourJob(stream);
+ *     // Here the lease ends and an event is recorded in `stream`.
+ *     // When GPU reaches this event, this worker object will be available for reuse
+ *     // on other streams for the same CUDA device.
+ *   }
+ * };
+ *
+ * ...
+ *
+ *   MyClass cls;
+ *   cls.DoSomeJob(stream1);  // create a new GPUWorker object and use it
+ *   cls.DoSomeJob(stream1);  // reuse the same object
+ *   cls.DoSomeJob(stream2);  // if the previous job has not finished, create a new GPUWorker
+ *   cls.DoSomeJob(stream2);  // reuse the second worker
+ *   cudaDeviceSynchronize();
+ *   cla.DoSomeJob(stream2);  // associated work is finished, reuses any of the two workers
+ * ```
+ *
+ * Possible future extensions:
+ *  - limit the per-device capacity and wait?
+ *  - user-provided functor to create new objects, if none are available?
+ *
+ * @tparam T                              type of the managed object.
+ * @tparam mutex_type                     type of the synchronization object used to protect
+ *                                        internal structures
+ * @tparam reuse_pending_on_same_stream   if true, subsequent requests for an object
+ *                                        specifying same stream may return an object associated
+ *                                        with an unfinished job on that stream
+ */
+template <typename T, typename mutex_type = std::mutex, bool reuse_pending_on_same_stream = true>
+class PerStreamPool {
+ private:
+  struct ListNode;
+  using ListNodeUPtr = std::unique_ptr<ListNode>;
+  struct ListNode {
+    T object;
+    int device_id = -1;
+    cudaStream_t stream = nullptr;
+    CUDAEvent event;
+    ListNodeUPtr next;
+
+    template <typename... ObjectConstructorArgs>
+    ListNode(int device_id, cudaStream_t stream, ObjectConstructorArgs&&... args)
+    : object(std::forward<ObjectConstructorArgs>(args)...)
+    , device_id(device_id)
+    , stream(stream)
+    , event(CUDAEvent::Create()) {
+    }
+
+    ~ListNode() {
+      event.reset();
```

Comment:
Do you need to call `reset` here?

Reply:
It's not necessary, but it doesn't hurt either. It's a leftover from a prototype where I didn't used `CUDAStream`. I'll remove it if there are more serious problems.